### PR TITLE
[WIP]Support moving table between different dbs/servers

### PR DESCRIPTION
**Pull request by everpcpc**
It is based on https://github.com/github/gh-ost/pull/557
Related issue: https://github.com/github/gh-ost/issues/417

### Description

This PR adds support for moving a table between different databases as well as different servers.

It's a feature complete version. And I'm working on tests/docs.

### CODE REVIEW

**Comment by tomkrouper** [comment]
This is super exciting. Thanks for working on this.

**Comment by shlomi-noach** [comment]
Wow, thank you!

Please note it will probably take me a few weeks, due to my timetable, to fully review this.

File: go/base/context.go [hunk 1]
```diff
@@ -190,6 +204,7 @@ type MigrationContext struct {
 	SharedColumns                    *sql.ColumnList
 	ColumnRenameMap                  map[string]string
 	DroppedColumnsMap                map[string]bool
+	SourceSchema                     string
```
Comment (MOON-CLJ) [inline go/base/context.go]: SourceSchema not used?

File: go/base/context.go [hunk 2]
```diff
@@ -279,49 +297,59 @@ func (this *MigrationContext) GetOldTableName() string {
...
 // same database instance. This would be true when running directly on master or when
 // testing on replica.
-func (this *MigrationContext) InspectorIsAlsoApplier() bool {
-	return this.InspectorConnectionConfig.Equals(this.ApplierConnectionConfig)
+func (this *MigrationContext) SrcIsAlsoDst() bool {
+	return this.DstMasterConnectionConfig.Equals(this.SrcMasterConnectionConfig)
+}
+
+// SrcMasterIsAlsoReplica is `true` when the master and replica is the same on source
+func (this *MigrationContext) SrcMasterIsAlsoReplica() bool {
```
Comment (MOON-CLJ) [inline go/base/context.go]: name  SrcMasterIsAlsoReplica and DstMasterIsAlsoReplica are very strange，and "the master and replica is the same on source" not make sense。

Comment (MOON-CLJ) [reply go/base/context.go]: but names "src srcMaster dst dstMaster" it's clear，i dont have  better idea。

File: go/logic/applier.go [hunk 3]
```diff
@@ -52,61 +53,82 @@ func newDmlBuildResultError(err error) *dmlBuildResult {
...
+		migrationContext:    migrationContext,
+		finishedMigrating:   0,
 	}
 }
 
 func (this *Applier) InitDBConnections() (err error) {
 
-	applierUri := this.connectionConfig.GetDBUri(this.migrationContext.DatabaseName)
-	if this.db, _, err = mysql.GetDB(this.migrationContext.Uuid, applierUri); err != nil {
+	srcApplierURI := this.srcConnectionConfig.GetDBUri(this.migrationContext.SrcDatabaseName)
```
Comment (MOON-CLJ) [inline go/logic/applier.go]: srcApplierURI name should not contain "Applier"，confirm to Inspector  https://github.com/github/gh-ost/pull/565/files#diff-5aa50dde7ca1baf7a819eecf740ea220R51

File: go/logic/applier.go [hunk 4]
```diff
@@ -211,86 +233,140 @@ func (this *Applier) CreateChangelogTable() error {
...
+		return err
+	}
+	log.Infof("Changelog table created on dst")
+
+	if this.migrationContext.SrcIsAlsoDst() {
+		return nil
+	}
+
+	log.Infof("Creating changelog table %s.%s on src",
+		sql.EscapeName(this.migrationContext.DstDatabaseName),
```
Comment (MOON-CLJ) [inline go/logic/applier.go]: this.migrationContext.DstDatabaseName seems should be this.migrationContext.SrcDatabaseName

**Pull request closed without merging**